### ISSUE TITLE
I want to use the wavelet scattering to fusion image, but I don't know how can I inverse transformation after fusion

### ISSUE THREAD

**Issue by strawberry1996** [kymatio/kymatio#301]
I want to use the wavelet scattering to fusion image, but I don't know how can I inverse transformation after fusion @

**Comment by lostanlen** [comment]
Hello! Welcome to Kymatio. It's unclear what you're asking. Could you please explain what you mean by fusion and inverse transformation?

**Comment by strawberry1996** [comment]
Thanks for your response,  I use wavelet to fusion the image before 
The algorithm of image fusion using DWT described in the following steps 
1. Size of inputs images: 
Given a two dimensional images (example, image A, image B) it is necessary to convert it into the same size a power of two square forms. 
2. Computation of two dimensions DWT: 
In this step, the two dimensional Discrete Wavelet Transform should be applied to the resized two dimensional images. 
3. Fusion rule: 
The most used of image fusion rule using wavelet transform is maximum selection, compare the two coefficients of DWT of the two images and select the maximum between. While the lowpass subband is an approximation of the input image, the three detail subbands convey information about the detail parts in horizontal, vertical and diagonal directions. Different merging procedures will be applied to approximation and detail subbands. Lowpass subband will be merged using simple averaging operations since they both contain approximations of the source images. 
4. Inverse discrete wavelet transforms: 
After selected the fused low frequency and high frequency bands, fused coefficient is reconstructed using the Inverse fast discrete wavelet transform to get the fused image which represent the new image.
 but now I want to use wavelet scatter instead of wavelet , I don't know if it will work.  and after image fusion, I need to do the inverse transformation to get a new picture.

**Comment by lostanlen** [comment]
Putting this into action does not require any change to the core of Kymatio, so I propose that we consider to close this issue, on the ground that the question of the OP has been answered.

@strawberry1996 Feel free to experiment with this on your own and please keep me updated about your progress.

**Comment by strawberry1996** [comment]
@lostanlen @eugenium 
I have tried it with the way which you tell me, but the result looks so terrible. I call the decode function of the reconstruction_exp.py. but it can't recovery the image. I think maybe I'm calling this function the wrong way. Could you help me to check it? Thanks very much. This is my call function:
 import os
import scipy

os.environ['CUDA_VISIBLE_DEVICES'] = "5"
import numpy as np
import nibabel as nib
import argparse
import sys
import torch
from PIL import Image
sys.path.append("/home/dengzeyu/python_code/run_on_server/pyscatwave-master")
# sys.path.append("F:\\project_python\\run_on_server\\pyscatwave-master")
import reconstruction_exp as recon
from scipy import misc

def parse_args():
    parser = argparse.ArgumentParser(description='fusion arguments')
    parser.add_argument('--M', type=int, default=128, help = 'height to resize')
    parser.add_argument('--N', type=int, default=128, help = 'width to resize')
    parser.add_argument('--J', type=int, default=2, help = 'size of scale')
    parser.add_argument('--cache', type=str, default='.cache', help = 'filename of cache for filter banks')
    args = parser.parse_args()
    return args

if __name__ == '__main__':
    args = parse_args()
    path = "/home/dengzeyu/python_code/run_on_server/pyscatwave-master/data/S1/"
    # path = "F:\\project_python\\run_on_server\\pyscatwave-master\\data\\S1\\"
    img = nib.load(path + "moment1.nii").get_data()
    im1 = np.zeros(shape=[1,3,128,128])
    im1[0, 0, :, :] = img
    im1[0, 1, :, :] = img
    im1[0, 2, :, :] = img
    # image = Image.fromarray(im1[0,0,:,:])
    # image.show()
    print(type(args))
    # print(args.N)
    # print(args.J)
    # print(args.cache)
    encoder = recon.ScatteringTranscoder(args.M, args.N, args.J)
    im2 = encoder.encode_processed(im1)
    im3 = encoder.decode(im2)
    print("im1: ", im1)
    im3Cpu = im3[0].cpu()
    image = im3Cpu.detach().numpy()
    image = image.squeeze()
    image = np.transpose(image, (1, 2, 0))
    print(type(im3[0]))
    print (im3[0].shape)
    misc.imsave("/home/dengzeyu/python_code/run_on_server/pyscatwave-master/test_data/outfile.jpg", image)

    print("im2: ", im2)

**Comment by strawberry1996** [comment]
OK, Thanks very much
----- 原始邮件 -----
发件人：Vincent Lostanlen <notifications@github.com>
收件人：kymatio/kymatio <kymatio@noreply.github.com>
抄送人：strawberry邓泽宇 <dengzeyu1996@sina.com>,  Mention <mention@noreply.github.com>
主题：Re: [kymatio/kymatio] I want to use the wavelet scattering to fusion image, but I don't know how can I inverse transformation after fusion (#301)
日期：2019年01月19日 01点31分

Putting this into action does not require any change to the core of Kymatio, so I propose that we consider to close this issue, on the ground that the question of the OP has been answered.
@strawberry1996 Feel free to experiment with this on your own and please keep me updated about your progress.

&mdash;

**Comment by eugenium** [comment]
You might want to try to increase the nubmer of iterations and potentially experiment with the learning rate in case that is the issue. It might however be just smoething with the scaling of the output or channel order.
Note that file also has a deprocess_img() function that might be helpful, you might want to try something like this

`
imgs = encoder.deprocess_img(im3)
imgs = imgs.clip(0, 1)
`

We'll need to add a comprehensive reconstruction examples at some point

**Comment by strawberry1996** [comment]
@eugenium  Hello, I want to know how to implement scattering inverse transform in decode function of the reconstruction_exp.py. I find that you through compute the loss between original image and reconstructed image to inverse transform. but I can't understand this "S_reconstructed = scattering(reconstructed, self.Psi, self.Phi, self.J)" , why are you get reconstructed image through the scattering? and I find there have three low pass filter in scattering function, I wonder what their respective roles are. Thank you for your patience

**Comment by eugenium** [comment]
reconstructed is the variable which is optimized over.  It's initialized randomly and then the following problem is solved:
\argmin_{reconstructed} || scattering(reconstructed) - S_imgs||_1

**Comment by strawberry1996** [comment]
@eugenium it is my understanding that randomly generate a variable, then use the variable to scattering.  then we use the scattering result and S_imgs to compute the loss.

**Comment by lostanlen** [comment]
@strawberry1996 related: #268
For the moment you have to look at the code, or try it out on a Dirac if you want to be sure empirically.

I suggest to close this issue. It is neither a feature request nor a bug report, but rather an open-ended research question about whether a certain computational pipeline (in this case, image fusion) can be developed with Kymatio. eugenium and eickenberg have replied to this question in the affirmative, and provided a starting point to address this research question. After five days of discussion, there does not appear to be an immediate need for a PR within the codebase Kymatio. Therefore, the status of this issue (open/closed) is largely unrelated to the state of the source tree. I don't want the discussions on the issue tracker to veer too far away of its main purpose, which is to develop and maintain the library. Rather, this sort of discussion is better targeted towards a discussion group, such as the PyTorch discussion group: https://discuss.pytorch.org/

**Comment by strawberry1996** [comment]
Thank you for your patience these days，I think I misunderstood the Issue's purpose. I'll take a hard look at the code. I'm sorry to bother you (。・＿・。)
